sklearn/ensemble/_hist_gradient_boosting/_predictor.pyx

```diff
@@ -21,85 +21,168 @@ from .common cimport node_struct
+        """Get reference to node."""
+        return &self.nodes[node_idx]
+
+    cdef int get_n_nodes(self) nogil:
+        """Get number of nodes in tree."""
+        return self.nodes.size()
+
+    def get_n_leaf_nodes(self):
+        """Return number of leaves."""
+        cdef node_struct node
+        return sum(node.is_leaf for node in self.nodes)
+
+    def get_max_depth(self):
+        """Return maximum depth among all leaves."""
+        cdef node_struct node
+        return max(node.depth for node in self.nodes)
+
+    def predict(self, const X_DTYPE_C [:, :] X):
+        """Predict raw values for non-binned data.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            The input samples.
+
+        Returns
+        -------
+        y : ndarray, shape (n_samples,)
+            The raw predicted values.
+        """
+        cdef:
+            int i
+            Y_DTYPE_C [:] out = np.empty(X.shape[0], dtype=Y_DTYPE)
+
+        for i in prange(X.shape[0], schedule='static', nogil=True):
+            out[i] = _predict_one_from_numeric_data(self, X, i)
+        return np.asarray(out)
+
+    def predict_binned(self, const X_BINNED_DTYPE_C [:, :] X,
+                       const unsigned char missing_values_bin_idx):
+        """Predict raw values for binned data.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            The input samples.
+        missing_values_bin_idx : uint8
+            Index of the bin that is used for missing values. This is the
+            index of the last bin and is always equal to max_bins (as passed
+            to the GBDT classes), or equivalently to n_bins - 1.
+
+        Returns
+        -------
+        y : ndarray, shape (n_samples,)
+            The raw predicted values.
+        """
+        cdef:
+            int i
+            Y_DTYPE_C [:] out = np.empty(X.shape[0], dtype=Y_DTYPE)
+
+        for i in prange(X.shape[0], schedule='static', nogil=True):
+            out[i] = _predict_one_from_binned_data(self, X, i,
+                                                   missing_values_bin_idx)
+        return np.asarray(out)
+
+    def compute_partial_dependence(self, grid, target_features, out):
+        """Fast partial dependence computation.
+
+        Parameters
+        ----------
+        grid : ndarray, shape (n_samples, n_target_features)
+            The grid points on which the partial dependence should be
+            evaluated.
+        target_features : ndarray, shape (n_target_features)
+            The set of target features for which the partial dependence
+            should be evaluated.
+        out : ndarray, shape (n_samples)
+            The value of the partial dependence function on each grid
+            point.
+        """
+        _compute_partial_dependence(self, grid, target_features, out)
+
+    def _set_node(self, int node_idx, unsigned int left=0,
```

Comment:
Maybe we should instead define this in the test file where it's used?

Reply:
The access from python makes a copy. 

```py
%%cython
# distutils: language=c++

from libcpp.vector cimport vector

cdef struct inner:
    int num

cdef class Container:
    cdef public vector[inner] items
    
        
def make_container():
    cdef:
        Container obj = Container()
        
    obj.items.resize(2)
    obj.items[0].num = 4
    obj.items[1].num = 10
    return obj
```

```py
obj = make_container()
obj.items
# [{'num': 4}, {'num': 10}]
obj.items[0]['num'] = 20
obj.items
# [{'num': 4}, {'num': 10}]
```

Reply:
Hm, I find this quite unexpected... is this documented in Cython anywhere?

Reply:
https://cython.readthedocs.io/en/latest/src/userguide/wrapping_CPlusPlus.html#standard-library

> All conversions create a new container and copy the data into it. The items in the containers are converted to a corresponding type automatically, which includes recursively converting containers inside of containers, e.g. a C++ vector of maps of strings.

Reply:
ok, thanks. That means we need to be very careful about any non-Cython access. Let's add a comment in the .pxd file linking to the docs you mentionned?

Let's also add a comment indicating we need this method here (in Cython) because Python access creates a copy